pkg/gatewayserver/io/ws/ws.go

```diff
@@ -226,9 +226,26 @@ func (s *srv) handleTraffic(c echo.Context) (err error) {
 		err = nil // Errors are sent over the websocket connection that is established by this point.
 	}()
 
+	// pingTicker is used to send periodic pings to the gateway.
+	// Ping direction; Server --> Gateway.
 	pingTicker := time.NewTicker(s.cfg.WSPingInterval)
 	defer pingTicker.Stop()
 
+	// pongTimeoutTicker is used to check when gateways don't send a pong to the server ping.
+	pongTimeoutTicker := NewCustomTicker(s.cfg.WSPingInterval * 2)
```

Comment:
The ticker interval `s.cfg.WSPingInterval * 2` is a magic number based on the fact that we should disconnect only after missing 2 pings. We could make this configurable if necessary.

Reply:
The handlers are within the same goroutine (they get executed in `ReadMessage`) so you can actually just use a counter and the existing ping ticker.

Would the following also work ?
A boolean that maintains if we ever saw a ping/ping pong from the gateway, and a counter of missed ping/pongs. If the gateway exceeds the limit of ping/pongs, and we did see it respond to them in the past, disconnect.

Reply:
> The handlers are within the same goroutine (they get executed in `ReadMessage`) so you can actually just use a counter and the existing ping ticker.
> 
> Would the following also work ? A boolean that maintains if we ever saw a pong from the gateway, and a counter of missed pongs. If the gateway exceeds the limit of pongs, and we did see it respond to them in the past, disconnect on the next ping. On pong, we reset the fault counter.

I also think this is a simpler way and this is also how I initially envisaged this to work.